### ISSUE TITLE
A problem on autoanchors with tensor.shape

### ISSUE THREAD

**Issue by GaryXXD** [ultralytics/yolov5#2394]
## ❔Question
Hi, I'm a rookie in object detection, I setup yolov5 environment completely and yolov5 can run successfully on dataset COCO128. (train, test, detect)
When I use my own dataset, a problem occurred and I don't know why and how to solve it.
My command setting is below:
Picture size are all 640x480 (original size is 4000x3000).
Model config is copied from yolov5s.yaml, but the number of classes is 89
python train.py 
--weights "none" 
--cfg "copy_from_5s.yaml" 
--data "my_dataset.yaml" 
--img_size 640 640 
--name "my_dataset_test"
The settings which aren't desplay here are default.

<img width="689" alt="螢幕擷取畫面 2021-03-08 141941" src="https://user-images.githubusercontent.com/19710955/110282618-c1451580-8019-11eb-9a2b-cdc1282b87db.png">

Please help me, thank you very much.

**Comment by glenn-jocher** [comment]
@GaryXXD 👋 Hello, thank you for your interest in our work! This issue seems to lack the minimum requirements for a proper response, or is insufficiently detailed for us to help you. Please note that most technical problems are due to:

- **Your modified or out-of-date code**. If your issue is not reproducible in a new `git clone` version of this repo we can not debug it. Before going further run this code and verify your issue persists:
```bash
$ git clone https://github.com/ultralytics/yolov5 yolov5_new  # clone latest
$ cd yolov5_new
$ python detect.py  # verify detection

# CODE TO REPRODUCE YOUR ISSUE HERE
```

- **Your custom data.** If your issue is not reproducible in one of our 3 common datasets ([COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [COCO128](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), or [VOC](https://github.com/ultralytics/yolov5/blob/master/data/voc.yaml)) we can not debug it. Visit our [Custom Training Tutorial](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data) for guidelines on training your custom data. Examine `train_batch0.jpg` and `test_batch0.jpg` for a sanity check of your labels and images.

- **Your environment.** If your issue is not reproducible in one of the verified environments below we can not debug it. If you are running YOLOv5 locally, verify your environment meets all of the [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies specified below. If in doubt, download Python 3.8.0 from https://www.python.org/, create a new [venv](https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/), and install requirements.

If none of these apply to you, we suggest you close this issue and raise a new one using the 🐛 **Bug Report template**, providing screenshots and a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) of your issue. Thank you!


## Requirements

Python 3.8 or later with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies installed, including `torch>=1.7`. To install run:
```bash
$ pip install -r requirements.txt
```


## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab and Kaggle** notebooks with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>


## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are passing. These tests evaluate proper operation of basic YOLOv5 functionality, including training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([test.py](https://github.com/ultralytics/yolov5/blob/master/test.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/models/export.py)) on MacOS, Windows, and Ubuntu.

**Comment by GaryXXD** [comment]
@glenn-jocher Thanks for your response. I clone yoloV5 and check my environment again, but this problem still happens.
Sadly, I can't reproduce the problem with dataset you mentioned. 
So, I trace the code in utils autoanchor.py and find out the problem is probably on scipy kmeans function.

In utils.antoanchor, function kmean_anchors, row 121, the line of code is below

k, dist = kmeans(wh / s, n, iter=30)  # points, mean distance

the scipy kmeans function cause the problem, 
e.g. In my case, probably because of lack of my own data, the labels are less various. 
When run the code above, n=9, the function output k only contains 6 points.
This situation may cause the code in utils.antoanchor, row 49, which is below,

m.anchor_grid[:] = new_anchors.clone().view_as(m.anchor_grid)  # for inference

to make mistake on reshaping from tensor(12 elements) to tensor(18 elements).
I'm not sure whether it is bug or not.

Thanks for spending time reading this

**Comment by glenn-jocher** [comment]
@GaryXXD oh, thanks for the details, interesting edge case failure mode. 

If there are 9 points requested but the scipy kmeans function only returns 6 points then yes autoanchor will break. Can you submit your labels.jpg and labels_correlogram.jpg? How many instances does your dataset have? Are most instances the same shape?

**Comment by glenn-jocher** [comment]
@GaryXXD maybe we could add an assert to at least better clarify the origin of the error:
```python
    s = wh.std(0)  # sigmas for whitening
    k, dist = kmeans(wh / s, n, iter=30)  # points, mean distance
    assert len(k) == n, print(f'{prefix}ERROR: scipy.cluster.vq.kmeans requested {n} points but returned only {len(k)}')
```

There is no good solution, as an incident like this is mainly an indicator of underlying data issues. The data appears suitable to the number of anchors requested, and the current architecture (3 or 4 output layers) may not be appropriate for monotonically sized data.

The only real quick fix available is to skip autoanchor with `python train.py --noautoanchor`

**Comment by GaryXXD** [comment]
@glenn-jocher Thanks for your response again. My label examples are below,
[concat_outcome.txt](https://github.com/ultralytics/yolov5/files/6138450/concat_outcome.txt)

There are 350 images, each image contains one object. 
These images are about screen dialoguebox detect, some of them are similar.
I think it is the point which you mentioned.
Thanks for your help.

**Comment by glenn-jocher** [comment]
@GaryXXD I've created a PR that should solve this in https://github.com/ultralytics/yolov5/pull/2470

Can you clone this branch and see if your issue is fixed with it?
```bash
git clone https://github.com/ultralytics/yolov5 -b glenn-jocher-patch-1  # clone PR branch
```

**Comment by GaryXXD** [comment]
@glenn-jocher It works. Now, it automatically skip anchor calcutlation if the length is not match.
![image](https://user-images.githubusercontent.com/19710955/111099874-68c5c900-8581-11eb-9d2f-f31c53a315b0.png)

Thanks a lot.

**Comment by glenn-jocher** [comment]
@GaryXXD ok, well that's an improvement at least. Though I see your BPR is only 0.65, which means that your recall and mAP will never be higher than 0.65. You should try to supply your own manual anchors in your yolov5s.yaml file if possible to better match your data:

https://github.com/ultralytics/yolov5/blob/38ff499b26b9f8bf183cd1c08746dd33d000eb59/models/yolov5s.yaml#L6-L11
